File: ios/App/Chat/Composer.swift
// ---------------------------------------------------------------------------
// Composer — the chat input dock, mirroring the Android Composer
// (android/.../chat/Composer.kt) and the webui Composer
// (gateway/webui/src/components/dock/composer.tsx).
//
// A paper-surface rounded card holding a text field over a button row: mic
// toggle, TTS toggle, a spacer, optional interrupt, and send. Send is disabled
// when the field is empty OR the SDK is not READY. Interrupt is shown only when
// cognition != .idle || isSpeaking (a cycle is in flight or audio is playing).
// TTS toggle flips the server-of-record preference via setTtsEnabled.
//
// Mic button (E5): tapping it gates on the AVAudio record permission. If already
// granted → toggle voice immediately; otherwise request it (iOS 17+
// AVAudioApplication.requestRecordPermission, AVAudioSession fallback) and toggle
// on grant; on denial show a one-shot inline notice and do NOT start (audio rule:
// graceful mic-denial fallback). When voiceMode .active the mic button wears the
// accent "mic-on" styling. An already-active mic stops without a permission check.
//
// The composer owns only the draft text + the mic-denied flag (local @State);
// everything else is read from SdkState and dispatched up through callbacks. The
// host docks it via .safeAreaInset(edge:.bottom); SwiftUI lifts it above the
// keyboard.
//
// accessibilityIdentifiers: chat-input, chat-send, chat-interrupt,
// chat-tts-toggle, chat-mic, mic-denied-notice.
// ---------------------------------------------------------------------------
import AVFoundation
import SwiftUI
import MobileSdk

struct Composer: View {
    /// True when the SDK is READY (text submission flows).
    let canSend: Bool
    /// Server-of-record TTS preference (mirrored, not owned).
    let ttsEnabled: Bool
    /// True while voiceMode == .active.
    let micActive: Bool
    /// True when a cycle is in flight or audio is playing.
    let canInterrupt: Bool
    let onSend: (String) -> Void
    let onMicToggle: () -> Void
    let onTtsToggle: () -> Void
    let onInterrupt: () -> Void

    @State private var draft = ""
    @State private var micDenied = false
    @FocusState private var inputFocused: Bool

    private let log = AppLog("composer")
    private static let micDeniedNotice = "Microphone access is needed for voice. Enable it in Settings."

    private var sendEnabled: Bool {
        !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && canSend
    }

    var body: some View {
        VStack(spacing: Space.sm) {
            if micDenied { micDeniedRow }
            draftField
            buttonRow
        }
        .padding(Space.md)
        .background(DuskColors.paper, in: RoundedRectangle(cornerRadius: ComposerLayout.radius))
        .shadow(
            // Soft amber halo — mirrors the webui composer glow; intensifies while
            // the mic is active (listening state). Color from DuskColors.amber token.
            color: DuskColors.amber.opacity(
                micActive ? ComposerLayout.glowListeningOpacity : ComposerLayout.glowOpacity
            ),
            radius: ComposerLayout.glowRadius
        )
        .overlay(
            // Listening glow: the whole composer card borders accent while the mic
            // is active (mirrors webui .composer--listening).
            RoundedRectangle(cornerRadius: ComposerLayout.radius)
                .stroke(micActive ? DuskColors.accent : DuskColors.line, lineWidth: 1)
        )
        .padding(.horizontal, Space.lg)
        .padding(.vertical, Space.md)
        .simultaneousGesture(
            // Swipe-down to dismiss keyboard; simultaneousGesture preserves TextField
            // touch handling (text selection / caret placement).
            DragGesture(minimumDistance: ComposerLayout.dismissDragThreshold)
                .onEnded { value in
                    if value.translation.height > ComposerLayout.dismissDragThreshold {
                        inputFocused = false
                        log.info("keyboard.dismiss reason=swipe-down")
                    }
                }
        )
    }

    // ── Derived state ───────────────────────────────────────────────────────

    private var showWave: Bool { micActive && draft.isEmpty }

    // ── Draft field ─────────────────────────────────────────────────────────

    private var draftField: some View {
        ZStack(alignment: .leading) {
            TextField(
                showWave ? "" : (canInterrupt ? "Type to interrupt…" : "Message Sentient"),
                text: $draft,
                axis: .vertical
            )
            .lineLimit(1...6)
            .font(Typo.ui(TypeScale.base))
            .foregroundStyle(DuskColors.ink)
            .tint(DuskColors.accent)
            .focused($inputFocused)
            .padding(.vertical, Space.xs)
            .accessibilityIdentifier("chat-input")
            if showWave { ListeningWaveform() }
        }
    }

    // ── Mic-denied notice ─────────────────────────────────────────────────────

    private var micDeniedRow: some View {
        Text(Self.micDeniedNotice)
            .font(.system(size: TypeScale.sm))
            .foregroundStyle(DuskColors.stop)
            .frame(maxWidth: .infinity, alignment: .leading)
            .accessibilityIdentifier("mic-denied-notice")
    }

    // ── Button row ──────────────────────────────────────────────────────────

    private var buttonRow: some View {
        HStack(spacing: Space.sm) {
            // mic + TTS are toggles: rounded-square, slashed glyph + sunk bg when
            // off, accent glyph + accent-tint bg + accent border when on (webui
            // icon-btn--mic-on/off, tts-on/off).
            ComposerToggle(systemName: micActive ? "mic" : "mic.slash", on: micActive, action: onMicTap)
                .accessibilityLabel("Microphone")
                .accessibilityIdentifier("chat-mic")
            ComposerToggle(
                systemName: ttsEnabled ? "speaker.wave.2" : "speaker.slash",
                on: ttsEnabled,
                action: onTtsToggle
            )
            .accessibilityLabel("Toggle speech")
            .accessibilityIdentifier("chat-tts-toggle")
            ComposerToggle(systemName: "paperclip", on: false, action: {})
                .accessibilityLabel("Attach")
                .accessibilityIdentifier("chat-attach")
            Spacer()
            if canInterrupt {
                Button(action: onInterrupt) {
                    RoundedRectangle(cornerRadius: ComposerLayout.stopGlyphRadius).fill(DuskColors.stop)
                        .frame(width: ComposerLayout.stopIconSize, height: ComposerLayout.stopIconSize)
                        .frame(width: ComposerLayout.buttonSize, height: ComposerLayout.buttonSize)
                        .background(DuskColors.stop.opacity(0.16), in: RoundedRectangle(cornerRadius: Radii.sm))
                        .overlay(RoundedRectangle(cornerRadius: Radii.sm).stroke(DuskColors.stop.opacity(0.35), lineWidth: 1))
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Stop")
                .accessibilityIdentifier("chat-interrupt")
            }
            ComposerAction(
                systemName: "paperplane.fill",
                tint: sendEnabled ? DuskColors.accent : DuskColors.ink4,
                action: submit
            )
            .disabled(!sendEnabled)
            .accessibilityLabel("Send")
            .accessibilityIdentifier("chat-send")
        }
    }

    // ── Submit ──────────────────────────────────────────────────────────────

    private func submit() {
        let trimmed = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, canSend else { return }
        onSend(trimmed)
        draft = ""
    }

    // ── Mic permission gate ───────────────────────────────────────────────────
    //
    // Active mic → stop without a permission check. Inactive → gate on the
    // AVAudio record permission: granted toggles immediately; undetermined
    // requests it and toggles on grant; denied shows the inline notice and does
    // NOT start. The request hop lands back on the main actor before mutating UI.

    private func onMicTap() {
        if micActive {
            onMicToggle()
            return
        }
        switch MicPermission.status() {
        case .granted:
            micDenied = false
            onMicToggle()
        case .denied:
            log.warn("micTap denied")
            micDenied = true
        case .undetermined:
            log.info("micTap requesting permission")
            MicPermission.request { granted in
                Task { @MainActor in
                    log.info("micPermissionResult granted=\(granted)")
                    if granted {
                        micDenied = false
                        onMicToggle()
                    } else {
                        micDenied = true
                    }
                }
            }
        }
    }
}

/// Record-permission gate over AVAudio. The app's deployment target is iOS 17,
/// so this routes through `AVAudioApplication` (the iOS 17+ replacement for the
/// deprecated `AVAudioSession` permission API). Three states mirror the Android
/// RECORD_AUDIO gate (granted / denied / undetermined).
private enum MicPermission {
    enum Status { case granted, denied, undetermined }

    static func status() -> Status {
        switch AVAudioApplication.shared.recordPermission {
        case .granted: return .granted
        case .denied: return .denied
        default: return .undetermined
        }
    }

    static func request(_ completion: @escaping (Bool) -> Void) {
        AVAudioApplication.requestRecordPermission(completionHandler: completion)
    }
}

#Preview("Default") {
    VStack {
        Spacer()
        Composer(
            canSend: true,
            ttsEnabled: true,
            micActive: false,
            canInterrupt: false,
            onSend: { _ in },
            onMicToggle: {},
            onTtsToggle: {},
            onInterrupt: {}
        )
    }
    .background(DuskColors.bg)
}

#Preview("Mic on (waveform + Listening…)") {
    VStack {
        Spacer()
        Composer(
            canSend: true,
            ttsEnabled: true,
            micActive: true,
            canInterrupt: false,
            onSend: { _ in },
            onMicToggle: {},
            onTtsToggle: {},
            onInterrupt: {}
        )
    }
    .background(DuskColors.bg)
}

#Preview("Streaming (Type to interrupt… + tinted stop)") {
    VStack {
        Spacer()
        Composer(
            canSend: true,
            ttsEnabled: true,
            micActive: false,
            canInterrupt: true,
            onSend: { _ in },
            onMicToggle: {},
            onTtsToggle: {},
            onInterrupt: {}
        )
    }
    .background(DuskColors.bg)
}

File: android/src/main/kotlin/io/sentient/android/chat/Composer.kt
// ---------------------------------------------------------------------------
// Composer — the chat input dock, mirroring the webui Composer
// (gateway/webui/src/components/dock/composer.tsx + components.css .composer).
//
// A paper-surface rounded card holding a multi-line text field over a button
// row: mic toggle, TTS toggle, a spacer, send, and (conditionally) interrupt.
// Send is enabled whenever the field is non-empty; a send issued before READY is
// queued by ChatScreen and flushed on the READY edge (always-typeable, web-sdk
// parity). canSend only tints the send glyph (accent=READY, muted=will queue).
// Interrupt is shown only when cognition != IDLE || isSpeaking (a cycle is in flight or
// audio is playing). Mic toggle flips startMic/stopMic — the full mic pipeline
// is Phase-3 (E3), so it just latches voiceMode for now. TTS toggle flips the
// server-of-record preference via setTtsEnabled.
//
// imePadding keeps the dock above the soft keyboard (spec §6.1). The composer
// owns only the draft text (local UI state); everything else is read from
// SdkState and dispatched up through the callbacks.
//
// Mic button (E5): tapping it gates on the RECORD_AUDIO runtime permission. If
// already granted → toggle voice immediately; otherwise launch the system
// prompt and toggle on grant; on denial show a one-shot inline notice and do
// NOT start (audio rule: graceful mic-denial fallback). When voiceMode ACTIVE
// the mic button wears the accent "mic-on" styling.
//
// testTags: chat-input, chat-send, chat-interrupt, chat-tts-toggle, chat-mic, chat-attach.
// ---------------------------------------------------------------------------
package io.sentient.android.chat

import android.Manifest
import android.content.pm.PackageManager
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import io.sentient.android.R
import io.sentient.android.theme.LocalTokens
import io.sentient.mobilesdk.design.Colors
import io.sentient.mobilesdk.log.createLogger

private val COMPOSER_RADIUS = 24.dp
private const val MIC_DENIED_NOTICE = "Microphone permission is needed for voice."
private val composerLog = createLogger("android", "composer")

/**
 * Stateless composer. The single mutation entry is the [onSend] callback; mic /
 * TTS / interrupt fire their own callbacks. Draft text is local UI state.
 *
 * @param canSend True when the SDK is READY (text submission flows).
 * @param ttsEnabled Server-of-record TTS preference (mirrored, not owned).
 * @param micActive True while voiceMode == ACTIVE.
 * @param canInterrupt True when a cycle is in flight or audio is playing.
 */
@Composable
fun Composer(
    canSend: Boolean,
    ttsEnabled: Boolean,
    micActive: Boolean,
    canInterrupt: Boolean,
    onSend: (String) -> Unit,
    onMicToggle: () -> Unit,
    onTtsToggle: () -> Unit,
    onInterrupt: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val tokens = LocalTokens.current
    val context = LocalContext.current
    var draft by remember { mutableStateOf("") }
    var micDenied by remember { mutableStateOf(false) }
    val sendEnabled = draft.trim().isNotEmpty()

    // RECORD_AUDIO runtime gate. On grant → toggle; on denial → inline notice,
    // do not start. Already-active mic stops without a permission check.
    val micLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission(),
    ) { granted ->
        composerLog.info("micPermissionResult", mapOf("granted" to granted))
        if (granted) {
            micDenied = false
            onMicToggle()
        } else {
            micDenied = true
        }
    }

    fun submit() {
        val trimmed = draft.trim()
        if (trimmed.isEmpty()) return
        onSend(trimmed)
        draft = ""
    }

    fun onMicTap() {
        if (micActive) {
            onMicToggle()
            return
        }
        val granted = ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) ==
            PackageManager.PERMISSION_GRANTED
        composerLog.info("micTap", mapOf("granted" to granted))
        if (granted) {
            micDenied = false
            onMicToggle()
        } else {
            micLauncher.launch(Manifest.permission.RECORD_AUDIO)
        }
    }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .imePadding()
            .padding(horizontal = tokens.space.lg, vertical = tokens.space.md)
            .clipCard(listening = micActive)
            .padding(tokens.space.md),
        verticalArrangement = Arrangement.spacedBy(tokens.space.sm),
    ) {
        if (micDenied) {
            Text(
                MIC_DENIED_NOTICE,
                color = Color(Colors.stop),
                fontSize = tokens.type.sm,
                modifier = Modifier.testTag("mic-denied-notice"),
            )
        }
        DraftField(
            draft = draft,
            micActive = micActive,
            streaming = canInterrupt,
            onChange = { draft = it },
            onSubmit = { submit() },
        )
        ButtonRow(
            sendEnabled = sendEnabled,
            canSend = canSend,
            ttsEnabled = ttsEnabled,
            micActive = micActive,
            canInterrupt = canInterrupt,
            onSend = { submit() },
            onMicTap = { onMicTap() },
            onTtsToggle = onTtsToggle,
            onInterrupt = onInterrupt,
        )
    }
}

/**
 * Paper card with the composer's rounded border. While [listening] (mic active)
 * the border glows accent — mirrors the webui .composer--listening state.
 */
private fun Modifier.clipCard(listening: Boolean): Modifier = this
    .background(Color(Colors.paper), RoundedCornerShape(COMPOSER_RADIUS))
    .border(
        1.dp,
        if (listening) Color(Colors.accent) else Color(Colors.line),
        RoundedCornerShape(COMPOSER_RADIUS),
    )

@Composable
private fun ButtonRow(
    sendEnabled: Boolean,
    canSend: Boolean,
    ttsEnabled: Boolean,
    micActive: Boolean,
    canInterrupt: Boolean,
    onSend: () -> Unit,
    onMicTap: () -> Unit,
    onTtsToggle: () -> Unit,
    onInterrupt: () -> Unit,
) {
    val tokens = LocalTokens.current
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(tokens.space.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // mic + TTS are toggles: slashed glyph + sunk bg when off, accent glyph +
        // accent-tint bg + accent border when on (webui icon-btn--mic-on/off, tts).
        ComposerToggle(
            iconRes = if (micActive) R.drawable.ic_mic else R.drawable.ic_mic_off,
            on = micActive,
            contentDescription = "Microphone",
            testTag = "chat-mic",
            onClick = onMicTap,
        )
        ComposerToggle(
            iconRes = if (ttsEnabled) R.drawable.ic_volume_2 else R.drawable.ic_volume_x,
            on = ttsEnabled,
            contentDescription = "Toggle speech",
            testTag = "chat-tts-toggle",
            onClick = onTtsToggle,
        )
        ComposerToggle(
            iconRes = R.drawable.ic_attach,
            on = false,
            contentDescription = "Attach",
            testTag = "chat-attach",
            onClick = {},
        )
        Row(modifier = Modifier.weight(1f)) {}
        if (canInterrupt) {
            Box(
                modifier = Modifier
                    .size(BUTTON_SIZE)
                    .clip(RoundedCornerShape(BUTTON_RADIUS))
                    .background(Color(Colors.stop).copy(alpha = 0.16f))
                    .border(1.dp, Color(Colors.stop).copy(alpha = 0.35f), RoundedCornerShape(BUTTON_RADIUS))
                    .clickable(onClick = onInterrupt)
                    .testTag("chat-interrupt"),
                contentAlignment = Alignment.Center,
            ) {
                Box(
                    Modifier
                        .size(11.dp)
                        .clip(RoundedCornerShape(2.dp))
                        .background(Color(Colors.stop)),
                )
            }
        }
        // Tint is accent when the field is non-empty AND the SDK is READY (send goes now).
        // Muted when the field is non-empty but not READY — the tap will queue, not drop.
        ComposerAction(
            iconRes = R.drawable.ic_send,
            tint = if (sendEnabled && canSend) Color(Colors.accent) else Color(Colors.ink4),
            contentDescription = "Send",
            testTag = "chat-send",
            enabled = sendEnabled,
            onClick = onSend,
        )
    }
}

